### PR TITLE
Desugar passes, and proper Sexp tests

### PR DESCRIPTION
**Pull request by mariari**
This PR does a few things

1. Splits the previous `Sexp.hs<Test>` into `Sexp.hs` and `Sexp/SimplifiedPasses.hs`
2. Defines proper tests for the sexp data structure in `Sexp.hs`
3.  Moves the real passes to `Desugared/Passes<Translate>`, and an export module `Desugar`
    - First we will copy the simplified passes then morph them to being fully proper tests.
4. Create tests for the real modules following the generated syntax in The `Doc/Architecture/Sexp.hs`

### CODE REVIEW

File: library/StandardLibrary/test/Sexp/SimplifiedPasses.hs [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+      recordWorksAsExpected,
+      sigandDefunWorksAsExpetcted,
+      moduleExpandsAsExpected
+    ]
+
+--------------------------------------------------------------------------------
+-- Simple Passes
+--------------------------------------------------------------------------------
+
+moduleTransform :: Sexp.T -> Sexp.T
```
Comment (cwgoes) [inline library/StandardLibrary/test/Sexp/SimplifiedPasses.hs:31]: Why does this exist in both the test file and the `Passes` file, shouldn't we be testing that one?

File: library/StandardLibrary/test/Sexp/SimplifiedPasses.hs [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+          | Sexp.isAtomNamed form "defun"
+              || Sexp.isAtomNamed form "type"
+              || Sexp.isAtomNamed form "defmodule"
+              || Sexp.isAtomNamed form "defsig",
+            Just name <- Sexp.atomFromT name =
+            name : acc
+        f _ acc = acc
+    moduleToRecord _ _ = error "malformed record"
+
+condTransform :: Sexp.T -> Sexp.T
```
Comment (cwgoes) [inline library/StandardLibrary/test/Sexp/SimplifiedPasses.hs:76]: Ditto

File: library/StandardLibrary/test/Sexp/SimplifiedPasses.hs [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+              |> Sexp.butLast
+       in Sexp.foldr generation acc (Sexp.butLast cdr)
+            |> Sexp.addMetaToCar atom
+    --
+    generation (Sexp.Cons condition body) acc =
+      Sexp.list [Sexp.atom "if", condition, Sexp.car body, acc]
+    generation _ _ =
+      error "malformed cond"
+
+ifTransform :: Sexp.T -> Sexp.T
```
Comment (cwgoes) [inline library/StandardLibrary/test/Sexp/SimplifiedPasses.hs:91]: Ditto

File: library/StandardLibrary/test/Sexp/SimplifiedPasses.hs [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+            |> Sexp.addMetaToCar atom
+    caseList pred then' =
+      [Sexp.atom "case", pred, Sexp.list [Sexp.atom "true", then']]
+    caseListElse pred then' else' =
+      caseList pred then' <> [Sexp.list [Sexp.atom "false", else']]
+
+-- This one and sig combining are odd mans out, as they happen on a
+-- list of transforms
+-- We will get rid of this as this should be the job of Code -> Context!
+multipleTransDefun :: [Sexp.T] -> [Sexp.T]
```
Comment (cwgoes) [inline library/StandardLibrary/test/Sexp/SimplifiedPasses.hs:111]: Ditto

File: library/StandardLibrary/test/Sexp/SimplifiedPasses.hs [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+         in combineMultiple name (defun : sameDefun)
+              |> Sexp.addMetaToCar a
+              |> (: search toSearch)
+    search (x : xs) = x : search xs
+    search [] = []
+
+-- This pass will also be removed, but is here for comparability
+-- reasons we just drop sigs with no defuns for now ☹. Fix this up when
+-- we remove this pass
+combineSig :: [Sexp.T] -> [Sexp.T]
```
Comment (cwgoes) [inline library/StandardLibrary/test/Sexp/SimplifiedPasses.hs:141]: Ditto and etc.

File: library/Translate/src/Juvix/Desugar/Passes.hs [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    --
+    generation (Sexp.Cons condition body) acc =
+      Sexp.list [Sexp.atom "if", condition, Sexp.car body, acc]
+    generation _ _ =
+      error "malformed cond"
+
+-- | @ifTransform@ - transforms a generated if form into a case
+-- - BNF input form:
+--   1. (if pred then else)
+--   2. (if pred then)
```
Comment (cwgoes) [inline library/Translate/src/Juvix/Desugar/Passes.hs:60]: when would we end up with this?

File: library/Translate/src/Juvix/Desugar/Passes.hs [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+          rest
+        ]
+        |> Sexp.addMetaToCar atom
+    moduleToRecord _ _ = error "malformed record"
+
+----------------------------------------
+-- Module Helpers
+----------------------------------------
+
+combine :: Sexp.T -> Sexp.T -> Sexp.T
```
Comment (cwgoes) [inline library/Translate/src/Juvix/Desugar/Passes.hs:326]: Needs comment at the top, what is this function doing?

File: library/Translate/src/Juvix/Desugar/Passes.hs [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+  | Sexp.isAtomNamed form ":defmodule" =
+    -- Turn this into a let-module
+    if  | Sexp.isAtomNamed (Sexp.car xs) ":cond" ->
+          Sexp.list [Sexp.atom ":let-mod", name, args, Sexp.car xs, expression]
+        | otherwise ->
+          Sexp.list [Sexp.atom ":let-mod", name, args, xs, expression]
+-- ignore other forms
+combine _ expression = expression
+
+ignoreCond :: Sexp.T -> (Sexp.T -> Sexp.T) -> Sexp.T
```
Comment (cwgoes) [inline library/Translate/src/Juvix/Desugar/Passes.hs:353]: Needs explanatory comment

File: library/Translate/src/Juvix/Desugar/Passes.hs [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+
+generatedRecord :: Sexp.T -> Sexp.T
+generatedRecord b =
+  Sexp.list
+    (Sexp.atom ":record" : fmap (\x -> Sexp.list [Sexp.Atom x]) (names b))
+
+names :: Sexp.T -> [Sexp.Atom]
+names body = Sexp.foldr grabNames [] body |> Set.fromList |> Set.toList
+
+grabNames :: Sexp.T -> [Sexp.Atom] -> [Sexp.Atom]
```
Comment (cwgoes) [inline library/Translate/src/Juvix/Desugar/Passes.hs:371]: Needs explanatory comment

File: library/Translate/src/Juvix/Desugar/Passes.hs [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+ignoreCond ((form Sexp.:> xs) Sexp.:> Sexp.Nil) trans
+  | Sexp.isAtomNamed form ":cond" =
+    Sexp.listStar [form, Sexp.foldr comb Sexp.Nil xs]
+  where
+    comb (pred Sexp.:> body) acc =
+      Sexp.listStar [Sexp.list [pred, trans body], acc]
+    comb _ acc = acc
+ignoreCond xs trans = trans xs
+
+generatedRecord :: Sexp.T -> Sexp.T
```
Comment (cwgoes) [inline library/Translate/src/Juvix/Desugar/Passes.hs:363]: Needs explanatory comment

File: library/Translate/test/Desugar/Sexp.hs [hunk 11]
```diff
@@ -0,0 +1,254 @@
+module Desugar.Sexp (top) where
+
+import qualified Juvix.Desugar.Passes as Desugar
```
Comment (cwgoes) [inline library/Translate/test/Desugar/Sexp.hs:3]: OK, so here we're testing these functions?

**Review by cwgoes — CHANGES_REQUESTED** [review]
Thanks; I'm a little confused about the testfiles, and a few nits, see comments.

File: library/StandardLibrary/test/Sexp/SimplifiedPasses.hs (hunk 1 above)
Comment (mariari) [reply library/StandardLibrary/test/Sexp/SimplifiedPasses.hs:31]: Simplified passes do not change with time, and are the simplified version of many pass types. I guess we can view it as a verified protoype. They also will keep passes we will remove from the frontend language etc etc.

File: library/Translate/src/Juvix/Desugar/Passes.hs (hunk 6 above)
Comment (mariari) [reply library/Translate/src/Juvix/Desugar/Passes.hs:60]: Every Cond.

```lisp
(cond
  ((even x) 3)
  (else     5)
```

File: library/Translate/test/Desugar/Sexp.hs (hunk 11 above)
Comment (mariari) [reply library/Translate/test/Desugar/Sexp.hs:3]: Yes

File: library/Translate/src/Juvix/Desugar/Passes.hs (hunk 7 above)
Comment (mariari) [reply library/Translate/src/Juvix/Desugar/Passes.hs:359]: done

File: library/Translate/src/Juvix/Desugar/Passes.hs (hunk 9 above)
Comment (mariari) [reply library/Translate/src/Juvix/Desugar/Passes.hs:409]: done

File: library/Translate/src/Juvix/Desugar/Passes.hs (hunk 8 above)
Comment (mariari) [reply library/Translate/src/Juvix/Desugar/Passes.hs:387]: done

File: library/Translate/src/Juvix/Desugar/Passes.hs (hunk 10 above)
Comment (mariari) [reply library/Translate/src/Juvix/Desugar/Passes.hs:398]: done

File: library/Translate/src/Juvix/Desugar/Passes.hs (hunk 6 above)
Comment (cwgoes) [reply library/Translate/src/Juvix/Desugar/Passes.hs:60]: Alright, thanks!

File: library/Translate/src/Juvix/Desugar/Passes.hs [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+        ]
+        |> Sexp.addMetaToCar atom
+    moduleToRecord _ _ = error "malformed record"
+
+----------------------------------------
+-- Module Helpers
+----------------------------------------
+
+-- | @combine@ - is the helper for transforming top level statements
+-- into expression
```
Comment (cwgoes) [inline library/Translate/src/Juvix/Desugar/Passes.hs:358]:
```suggestion
-- into expressions
```

**Review by cwgoes — APPROVED** [review]
ACK

**Pull request merged**